### DISCUSSION TITLE
Finding orphaned event handlers / hooks / functions

### DISCUSSION THREAD

**Discussion by vpax** [zeek/zeek#1970] [Ideas]
A common, not-so-easy-to-debug mistake in Zeek scripting is to use the wrong name in an event handler (or a hook), leading to it failing to get invoked.  For example:

> event zeke_init() { my_good_stuff(); }

will never run due to the typo in the name, but this is also not flagged.  Another version of this - more of a horror show - is:

> module MyModule;
> 
> global do_all_the_work: event();
> 
> event zeek_init() { event do_all_the_work(); }
> 
> event do_all_the_work() { print "did my work"; }

which does not in fact print `did my work` (it will though if the event is invoked as `MyModule::do_all_the_work`).

The branch `topic/vern/find-unused-work` has new functionality to analyze scripts to find such "orphaned" event handlers (and hooks).  It also finds functions that can't be called (they're not exported, and there's no path from something "live" that invokes them).

There are a couple of design questions:

1. Should Zeek always do this checking, or only if the user asks?  I worry that if it's optional, users often won't think to do the check when they're puzzling over why their script isn't working.  Turning it on all the time has a minor startup cost (about 50 msec on my MBP when compiled non-optimized), though it does also mean that we need to ensure that the base scripts and test suite don't generate any undue warnings (see next item).
2. What should we do about issues in the base script that are actually correct?  These arise when there _is_ a code path to a given event/hook/function, but it's not active by default - typically due to conditional code such as for cluster operation, or if the event is only instantiated via Broker.  My current solution for these is to annotate the event handler with `&is_used`. There are 14 such instances in the base scripts.

Finally, here are the orphans in the base scripts:

> frameworks/signatures/main.zeek: function has_signature_matched
> frameworks/software/main.zeek: function software_endpoint_name
> frameworks/notice/main.zeek: function execute_with_notice
> frameworks/sumstats/non-cluster.zeek: function request
> frameworks/sumstats/plugins/std-dev.zeek: hook std_dev_hook
> protocols/ftp/utils-commands.zeek: function pop_pending_cmd
> protocols/smb/smb1-main.zeek: event smb_pipe_bind_ack_response
> protocols/smb/smb1-main.zeek: event smb_pipe_bind_request
> protocols/smb/smb1-main.zeek: event smb_pipe_request

along with the `attr_types`, `nas_port_types`, `service_types` and `framed_protocol_types` globals in `protocols/radius/consts.zeek` (which get flagged because the analyzer spots their lambdas as orphaned, because the constants aren't exported and aren't used anywhere).

Currently I mark issues like these with `&deprecated` (which the analyzer then knows not to complain about).

If this design sounds good, I'll put together an associated PR.

**Comment by bbannier** [comment]
> What should we do about issues in the base script that are actually correct? These arise when there is a code path to a given event/hook/function, but it's not active by default - typically due to conditional code such as for cluster operation, or if the event is only instantiated via Broker.

I skimmed your branch and was wondering whether it would be possible to treat these different cases differently

- code is behind preprocessor conditionals: would it be possible to also conditionally define the "unused" functions/hooks/events?
- function passed into Broker: should this be a use for your analysis?
- unexported function used in module extension not loaded by default: would these go away if the "unused" functions were moved to the files which actually use them?

I haven't tried to make these changes, but at least on the surface not having to annotate them with `&is_used` would seem cleaner to me. The attribute looks more like an escape hatch which could just as well hide issues (e.g., refactorings gone wrong).

> Should Zeek always do this checking, or only if the user asks? I worry that if it's optional, users often won't think to do the check when they're puzzling over why their script isn't working.

Currently deprecations trigger warnings and then fail when the deprecated code was removed. I could imagine that one could similarly always triggering warnings on unreachable code, but also add a new flag which turns any warnings into hard errors (I am probably thinking of something along the lines of `-Wdeprecated`, `-Wusage-isses`, `-Wunreachable-...`, matching `-Wno-deprecated` and a global `-Werror` which possibly takes args, e.g., `-Werror=deprecated`, ...). This probably applies to many analyses from `AnalOpt`.

At least most of the compiled languages I am familiar with seem to follow such a pattern of some level warning by default and a way of turning warnings into errors. If warnings are not enabled by default this usually leads to messy code (e.g., Bison, or C and C++ compilers defaulting to no warnings by default). Warnings about unused code appear to me to be more in the useful diagnostics and less in the linter camp, and I would expect something like this to be on by default, but probably not lead to fatal errors.

I am also wondering how useful it would be to make such an attribute available to user code. Allowing to suppress warnings on the source level may make it harder to spot issues; having suppressions on the zeek invocation would make it clearer what is allowed. If users rely on many scripts they do not control (e.g., installed via `zkg`) they might need to suppress many warnings globally, but wouldn't be able to annotate the function declarations, so the attribute wouldn't help them there much either.

**Reply by vpax** [reply]
> * code is behind preprocessor conditionals: would it be possible to also conditionally define the "unused" functions/hooks/events?

It's not generally possible.  The problem is that one `@if` idiom is to use it to conditionally load another script, like the cluster version versus the non-cluster version.  (See for example `base/frameworks/intel/__load__.zeek`.)  It would be possible to _partially_ analyze conditionals, by treating any potential identifier seen inside skipped-due-to-conditional text as "live".  That would probably work okay for some simpler conditional usages.  In principle it could introduce false negatives, but that seems very unlikely.

**Reply by vpax** [reply]
> * function passed into Broker: should this be a use for your analysis?

Those actually _are_ used in the analysis.  The problem arises instead when there are two completely separate scripts (that work together), and one of them fields some events only sent by the other.  I believe I only saw this style in the test suite, not in the base scripts, so it seems to me a bit less of a drag to fix those using `&is_used` annotations.

**Reply by vpax** [reply]
> * unexported function used in module extension not loaded by default: would these go away if the "unused" functions were moved to the files which actually use them?

Yes, that would address it.  That said, I don't believe I've encountered these.  The orphaned functions I flag looked vestigial to me (but would of course be useful to know if that's wrong).

**Reply by vpax** [reply]
> The attribute looks more like an escape hatch which could just as well hide issues (e.g., refactorings gone wrong).

One plus though is it's easy at any point to find instances of the attribute and assess them for whether they're still apt.  I think the current situation is decidedly worse, which is that a refactoring can silently mask that some functionality has become broken.

**Reply by vpax** [reply]
> I could imagine that one could similarly always triggering warnings on unreachable code, but also add a new flag which turns any warnings into hard errors

Sounds good to me.

**Reply by vpax** [reply]
> am also wondering how useful it would be to make such an attribute available to user code. Allowing to suppress warnings on the source level may make it harder to spot issues; having suppressions on the zeek invocation would make it clearer what is allowed.

This I'm not getting.  If we always (by default) generate warnings, surely we want the base scripts to not generate any warnings; given that, then we need the attribute.  Once the attribute is available in the scripting language, it's available to user code, since there's no distinction in Zeek scripting between internal scripting vs. external.

**Reply by vpax** [reply]
> If users rely on many scripts they do not control (e.g., installed via `zkg`) they might need to suppress many warnings globally, but wouldn't be able to annotate the function declarations, so the attribute wouldn't help them there much either.

Wouldn't this burden be on the package developer?  They shouldn't publish a package that generates warnings; they should use annotations if required to suppress those.

**Reply by bbannier** [reply]
> > am also wondering how useful it would be to make such an attribute available to user code. Allowing to suppress warnings on the source level may make it harder to spot issues; having suppressions on the zeek invocation would make it clearer what is allowed.
> 
> This I'm not getting. If we always (by default) generate warnings, surely we want the base scripts to not generate any warnings; given that, then we need the attribute. Once the attribute is available in the scripting language, it's available to user code, since there's no distinction in Zeek scripting between internal scripting vs. external.

Yes, the base scripts must not trigger any warnings since users do not control them. They should either be fixed or suppress warnings.

What I meant here is that if _only a command line flag_ would allow suppressing warnings it would be clear from the `zeek` invocation whether any used non-base script sources are "clean"; allowing suppression only via an attribute would always require manual auditing of the sources. Maybe the right approach lies somewhere in the middle, see below.

> > If users rely on many scripts they do not control (e.g., installed via `zkg`) they might need to suppress many warnings globally, but wouldn't be able to annotate the function declarations, so the attribute wouldn't help them there much either.

> Wouldn't this burden be on the package developer? They shouldn't publish a package that generates warnings; they should use annotations if required to suppress those.

The external package might not have produced any warnings when it was published, but Zeek might have become stricter in the meantime and new warnings are emitted now for users of that newer Zeek version. Users of the package will now see warnings from code they do not control until the package maintainer decides to fix it (if at all). These warnings could drown out warnings in code they actually do control, making it harder to e.g., write CI ensuring no warnings.

Coming back to the different suppression mechanisms, maybe one more mechanism could help here, e.g., one similar to C pragmas

```zeek
# strict_script.zeek

# Hypothetical attribute `&nowarn` suppresses all warnings from loaded
# file and probably its recursive dependencies. Maybe there's a better
# syntax possible which uses some sort of scope or pragma-like stack
# (push nowarn(unused), pop).
@load unmaintained/package &nowarn=(unused).

# File itself can use strict warning settings.
```

The base scripts could use a similar mechanism to ensure no warnings leak from essential input files, so the per-function/event/hook attribute wouldn't strictly be needed anymore.

**Reply by vpax** [reply]
> What I meant here is that if _only a command line flag_ would allow suppressing warnings it would be clear from the `zeek` invocation whether any used non-base script sources are "clean"; allowing suppression only via an attribute would always require manual auditing of the sources. Maybe the right approach lies somewhere in the middle, see below

Ah okay then I think we're actually in alignment.  It will be valuable to both have an attribute for saying "don't warn about this, it's actually used" and a command-line flag for "suppress warnings about unused functions".

> @load unmaintained/package &nowarn=(unused)

Perhaps at some point, but I imagine that coming up with an agreed-upon design for something like this would take a good deal of discussion, so I'd want to decouple that from getting the current functionality into Zeek.

**Comment by ckreibich** [comment]
This is great to have. I've been bitten by mistyped event names many times. While working on the cluster management framework I've often found that I put together all of the eventing with Broker etc for a new feature, and give if a shot. If it works, great! If not, argh, start checking all the things that I might have screwed up: sometimes it's a typo, sometimes I forgot to tell Broker to publish an event, sometimes I simply forgot to write a corresponding handler on the receiving side. Debugging aids for any of these are welcome.

> Should Zeek always do this checking, or only if the user asks?

If feasible, I'd warn by default and provide a flag to suppress the warning, assuming this squeezes out sufficient performance gain. If we're only talking about 50ms, it's likely not a big deal ... we've not been tracking startup time via regression testing, but it would be good to do so — Tim was recently wondering whether startup had gotten noticeably slower, so I'm wondering whether we sometimes introduce bigger culprits than 50ms anyway.

> What should we do about issues in the base scripts?

I'm all for keeping this clean via whatever mechanism we find suitable. Attribute-tagging seems a fine solution for the tricky cases.

If your use detector can easily be sufficiently granular, I wonder if we could skip it for the base scripts (given that we'd keep those tidy), and only run it on other scripts, such as those coming in via `zkg` or `local.zeek`. We've in the past discussed the ability for Zeek to tell when it's loading a package's scripts — that ability could come in handy for this.

I also think that if we can put a test that looks for this problem in the default testsuite for new Zeek packages, it's very likely to survive there.

**Reply by vpax** [reply]
@ckreibich thanks for the thoughts.

> If your use detector can easily be sufficiently granular, I wonder if we could skip it for the base scripts

Are you thinking of doing this to avoid nuisance warnings (doesn't quite fit with your parenthetical comment about tidiness)?  Or processing burden?  For the latter, I don't think that would be viable since it may well be that it's the base scripts that generate the event/hook/call that makes a user script live.

**Reply by vpax** [reply]
> I also think that if we can put a test that looks for this problem in the default testsuite for new Zeek packages, it's very likely to survive there.

I'm not following what you mean by "very likely to survive there" ... ?

**Reply by ckreibich** [reply]
> Are you thinking of doing this to avoid nuisance warnings?

It was to avoid processing burden, but — ah! — now I see that this doesn't apply, per your comment. 

> I'm not following what you mean by "very likely to survive there" ... ?

If a newly bootstrapped package automatically gets such a test case, it's relatively unlikely that package maintainers manually throw it out. I've noticed that any kind of test we put in there by default tends to stay, even when not particularly appropriate for the purpose of the package. I'm guessing it's because it looks daunting to modify, or simply because it's not strictly necessary to modify.

**Reply by vpax** [reply]
> If a newly bootstrapped package automatically gets such a test case, it's relatively unlikely that package maintainers manually throw it out. I've noticed that any kind of test we put in there by default tends to stay, even when not particularly appropriate for the purpose of the package. I'm guessing it's because it looks daunting to modify, or simply because it's not strictly necessary to modify.

Got it.  I see this working if we make the detection optional - we'd then add a test that turns it on, and that will fail if they've introduced an issue.  If though we make detection on-by-default (which I like, given the nature of this problem), with a flag required to turn it off, then this doesn't seem necessary since the package developer will be confronted with the issue during the development process.
